### DISCUSSION TITLE
Another Scratch Addons settings page redesign idea

### DISCUSSION THREAD

**Discussion by DNin01** [ScratchAddons/ScratchAddons#6317] [Discussion]
Here's another settings page redesign concept mockup I made.

<details open><summary>

### Main page

</summary>
<p>

![Scratch Addons settings redesign 2](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/7433823e-45c8-4031-8000-b31decc77d42)

The list of addons is on the left, and you can click an addon's name to customize it on the second column.

There are a bunch of things to notice there. All the addon's details like the description and even a screenshot are on that area, and the settings are right underneath them, no longer taking space away from the list of addons. Single-select settings are now dropdown menus and there's even a collapsible settings group, which has a slider setting! (This slider snaps, though, so no fine tuning.)

</p>
</details>

<details><summary>

### More pictures

</summary>
<p>

![Scratch Addons settings redesign 2 - explore tab](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/d056aac5-edad-4e9e-a906-a236eab81ed8)

</p>
</details> 

<details open><summary>This essentially implements all of these:</summary>
<p>

- #6352
- #6290
- #6009
- #5719
- #5320
- #3642

</p>
</details>

Overall, this is somewhat inspired by Microsoft PowerToys.

**Comment by WorldLanguages** [comment]
Each addon has its own icon. Very interesting!

Using a slider instead of a select dropdown (which is used for "animation speed" on this example) is also a nice idea.

**Reply by WorldLanguages** [reply]
Making a list of incremental changes we can make, without actually having to change the whole layout at once:

- Setting descriptions (there's an open PR already that suggests some descriptions)
- Related addons (there's an open PR implementation, maybe you could add this to your screenshot concept?)
- Using sliders for some select settings (such as four speed settings) and some number settings (#6319)
- Grouping addons by the area of Scratch they affect (for example, `Editor/Code`)
- Each addon already has its own screenshot on scratchaddons.com/addons

**Reply by WorldLanguages** [reply]
Forgot: having a small icon which identifies each addon!

**Reply by DNin01** [reply]
In my design, the category names would stick to the top of the list (like Sticky Scroll in VS Code) to replace scope-identifying icons.

**Reply by Secret-chest** [reply]
> Forgot: having a small icon which identifies each addon!

@WorldLanguages I'm open to draw them.

**Comment by WorldLanguages** [comment]
Would all addons be listed on the left side?
I was thinking we should only show enabled, recommended and featured addons. But possibly, other addons could be behind an "all addons" section, a "load more" button or a "show all addons" button.
Also, we probably still want to provide a list of all enabled addons, as well as a "new and updated" addon group. Maybe we can have a design similar to Scratch comment threads: show the first 3 enabled addons, and a "see more" button which gets transparent pixel-by-pixel (get what I mean?)

**Reply by DNin01** [reply]
Yeah, having so many addons can make the list harder to browse, so limiting the options by default might help prevent this list from becoming overwhelming. I find it difficult to think of an effective way to do that, though. Addon usefulness can be pretty opinionated.

**Reply by WorldLanguages** [reply]
@Secret-chest How would you solve this problem?

**Reply by WorldLanguages** [reply]
I think we must list enabled addons at the top of the page (except a small "new and updated addons" addon group which already goes below enabled addons today). I really think we have no other option, as going to the settings page to disable or change settings of an enabled addon is a typical user flow. As I said in another comment, grouping enabled addons by category/subcategory would also be helpful to the user.

@DNin01 Maybe at the top of your concept screenshot, we could have 3 buttons/sections: `Enabled addons`, `Explore addons` and `Scratch Addons settings`. The first two would be the same page, but they would scroll the user to the correct scroll point of the addon list. The "explore addons" button would scroll down the addon list right below the last enabled addon (remember I just said we should list all enabled addons at the top of the addon list). After the user clicked this button (or manually scrolled below enabled addons), we can display a few (disabled) addons per addon group, and provide a "show more/all" button for each addon group (when clicked, we would also show enabled addons within this group). This would, more or less, fix the 150+ addons issue, as users can still go through every addon by expanding every addon group, and we give them a chance to ignore subcategories they may not be interested about (like "forums" addons) by not clicking "show more" on those addon groups.

**Reply by DNin01** [reply]
This is looking pretty good already:

![Scratch Addons settings redesign 2 - explore tab](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/d056aac5-edad-4e9e-a906-a236eab81ed8)

**Reply by DNin01** [reply]
I would probably try to keep the "My addons" and "Explore addons" categories separate. "My addons" would be designed for customizing addons you enabled and would only support compact list view. "Explore addons" would be designed for browsing and finding addons to enable and would have three view options: grid (what you see in the picture), details (grid with pictures), and list (one-dimensional list with icons, names, and descriptions). One thing I didn't put on the picture was how we would open the settings and details for an addon from the explore page. Maybe a cog icon on each card and it would open the side pane?

**Reply by WorldLanguages** [reply]
> I would probably try to keep the "My addons" and "Explore addons" categories separate.

What do you mean by "separate"? I think they should be the same page in different scroll points.
If users don't explore this tab it's going to be problematic, even if we teach them how to use it the first time they visit the settings page, it's possible some users will forget about it. I'd prefer to sleep well by listing disabled addons ("explore addons") below the list of enabled ones.
They can be separate pages, but I believe we should add "infinite" scrolling (or call it whatever) to the "my addons / enabled addons" page, by listing disabled addons on it after the user has scrolled past all enabled addons.

**Reply by DNin01** [reply]
> > I would probably try to keep the "My addons" and "Explore addons" categories separate.
> 
> 
> 
> What do you mean by "separate"? I think they should be the same page in different scroll points.

By separate, I meant not connected in that you can't scroll from one to another.

**Reply by WorldLanguages** [reply]
> One thing I didn't put on the picture was how we would open the settings and details for an addon from the explore page. Maybe a cog icon on each card and it would open the side pane?

I like the idea of having "explore addons" take advantage of the whole screen, without needing a sidebar unless the user decided to enable an addon from there and tweak its settings. Maybe this sidebar could act as a modal (but still be positioned at the right border), this would mean you wouldn't be able to scroll through the "list of addons to explore" while changing the settings for the addon. Maybe this could also be the case on the "my addons" page?

**Reply by DNin01** [reply]
One way we could connect them without much effort would be to add a button to go to your enabled addons at the top of the Explore addons page and a button to go to the Explore addons page at the bottom of the page for the user's enabled addons.

I don't know how that sounds, though.

**Reply by DNin01** [reply]
> Maybe this sidebar could act as a modal (but still be positioned at the right border), this would mean you wouldn't be able to scroll through the "list of addons to explore" while changing the settings for the addon. Maybe this could also be the case on the "my addons" page?

Interesting. I was imagining that earlier.

Maybe some layouts could adapt to the side pane appearing (like the list layouts where the on/off switches would move out of the way) and in some layouts everything would stay in place when the modal is onscreen.

**Reply by WorldLanguages** [reply]
Dunno, I just think the "my addons" page is going to be the default tab when accessing the settings page, and if the user scrolls below all enabled addons, we have the opportunity to show them something, even though the user might not have clicked anything.

**Reply by WorldLanguages** [reply]
The "new and updated" addon group would probably stay on (and at the top of) the "my addons" page (as we want new addons to have the most visibility), so we can probably add a "explore more addons" button to the head/bottom of that addon group which links/scrolls to the list of disabled addons to explore. Makes it a bit less likely users won't find disabled addons.

**Reply by DNin01** [reply]
Not having as much information visible at once may even allow us to unload certain information such as non-details data (settings, CSS, JS, etc.) about disabled addons to cut down on memory usage. I know you talked about that in #5190. It wouldn't take long to load it again whenever the user opens it up in the details and settings pane.

**Reply by Hans5958** [reply]
I would rather to have ALL of it to show. We already have some on it on the settings page, so why bother filtering by default? The search has it's own purpose.

If anything, I would rather consolidate addons to make this simpler. I guess the problem of the small space is that we seemed to make the available settings too basic. WIth more space we can be more "elaborate" with these settings.

**Comment by WorldLanguages** [comment]
Would some "addon categories" (such as `Editor - General`) benefit if the addons on it were listed as a grid with screenshots (similarly to scratchaddons.com/addons) instead of using small icons that identify each addon?
I'm thinking we could maybe use rows vs. grids depending on the addon group/category. For example, costume editor addons would probably look better if each addon had a screenshot right in the addon list. Think 2D color picker, or onion skinning.

**Reply by WorldLanguages** [reply]
#5893 calls this "compact" view vs. "default" view.

Showing screenshots for addon groups such as "enabled addons" makes no sense. Meanwhile, some addon groups/categories will feel better if each addon has an accompanying screenshot.

**Reply by DNin01** [reply]
I don't know if I like the sound of different categories using different layouts, but I do think different layouts is worth thinking about.

Here are some ideas for potential layouts:
- A list view with addon icons and names (compact)
- A list view with addon icons, names, and descriptions (expanded)
- A grid view with addon screenshots, names, and descriptions (details)
- A grid view with addon icons, names, and descriptions (explore)

I like the idea of having options to switch between these, but keep in mind that when the list and the addon details and settings area are put side by side, these layouts may not have enough space to work as well. We could make that area smaller or move it to the bottom and it could work alright, though. There are many layout possibilities that come to mind when the list makes full use of the space. Having a details and settings column limits our options.

Maybe the details and settings area could act as a side pane that starts out hidden when you're just browsing the list of addons, and when you click on an addon for more information, it opens up and the addon cards in the list shrink and rearrange into a list like in my mockup so you can still see several of them on a single screen. When we don't have a side pane to worry about, more layout options are unlocked.

**Reply by WorldLanguages** [reply]
As a small conclusion, it would be nice if every addon had a screenshot and an icon which identified it.

**Reply by WorldLanguages** [reply]
Also, let's not forget the popup view of the settings page would continue to look more or less like it does now: no side-panel, as there's no space.

**Reply by DNin01** [reply]
> Also, let's not forget the popup view of the settings page would continue to look more or less like it does now: no side-panel, as there's no space.

It can probably just list enabled addons and potentially recently used ones and let you toggle them on and off. If we really wanted, we could add a settings button to each one that would open a popup inside the popup to change the settings. You wouldn't be able to see the list of addons behind it.

**Reply by WorldLanguages** [reply]
We do want to continue to support changing settings of enabled and searched addons through the popup. The UI doesn't really matter as long as it's possible to do so. We also should show addon descriptions, notices, and presets on this view, if the user clicks the addon or enables it.

**Reply by Secret-chest** [reply]
The #5893 allows the user to choose what's best for them. I don't like using a grid since it looks like a shopping site and makes scanning the list more difficult

**Comment by WorldLanguages** [comment]
We could add some buttons next (or below) the search bar that act as search filters. They would replace the current website/editor/popup filter, and we could also allow users to list addons that were assigned a certain tag (excluding tags that refer to a certain piece of Scratch, as those already have their own "addon groups")

- "Editor" (only editor addons - would still show subgroups when showing results)
- "Website" (only website addons - would still show subgroups when showing results)
- "Project players" (only project player addons)
- "Extension popup" (only extension popup addons)
- "Colors"
- "Productivity"
- "Accessibility"
- "Customization"
- "Beta" (?)

YouTube already does this, more or less. Note that there's buttons to scroll horizontally and show more search terms.
![image](https://github.com/ScratchAddons/ScratchAddons/assets/17484114/99024fef-0b5d-40a3-ac4d-3c50856b7a0e)

**Reply by Secret-chest** [reply]
I was thinking of special syntax but there could be buttons to insert it.

**Reply by WorldLanguages** [reply]
"special"? @Secret-chest

**Reply by Secret-chest** [reply]
> "special"? @Secret-chest 

My Samsung Galaxy A20e ate one word, oops @WorldLanguages

**Reply by WorldLanguages** [reply]
Yeah, these buttons could add special syntax to the search bar. It's a possibility.

**Reply by DNin01** [reply]
> I was thinking of special syntax but there could be buttons to insert it.

Making the user enter qualifiers in the search bar would not be as fast or easy to learn as buttons to apply filters.

**Reply by Secret-chest** [reply]
The buttons to enter these would still exist, but there will be more - like author:Secret-chest or version:1.7.0

**Comment by WorldLanguages** [comment]
We could possibly group enabled addons by category (which area of Scratch they affect) at the top of the page. The user might want to go to the settings page to check all of the addons they have enabled on, let's say, the costume editor, to disable one of them they don't remember the name of.

This would mean that when the user scrolls below the enabled addons, we can show three or so disabled addons per each group before having the user click "load more" or somehow interact with the addon group to see more addons on that group. Since enabled addons are already grouped above, we can move enabled addons to the bottom (or at least, "below the fold") on this view, and give the user a chance to read about addons they haven't enabled from that addon group.
At some point, we can try to be more clever about which addons we show without requiring the user to scroll or click a button. For example, remember that the user already tried out an addon and decided to disable it afterwards. I talk more about this on #5431 ("knowing whether a user recently tested out an addon can be valuable information at the time of considering what addons to suggest to the user"

**Comment by Hans5958** [comment]
We're getting closer, folks...

![image](https://github.com/ScratchAddons/ScratchAddons/assets/11584103/974a1ae0-830a-42a3-ab23-7123ce87345f)

**Comment by TheColaber** [comment]
![image](https://github.com/ScratchAddons/ScratchAddons/assets/72760579/11a4f885-02f6-4de9-83a7-7b40b8a63526)
here is a little something we talked about in discord

**Reply by Hans5958** [reply]
If there's anything about the settings idea, it would be to discard the thick paddings in favor of, no boxes at all.

(Also, please make another post. We don't have time to check the replies for other ideas.)

**Comment by Hans5958** [comment]
As another reference, here's the OG, Mega Scratch Userscript.

![image](https://github.com/ScratchAddons/ScratchAddons/assets/11584103/d571e9e5-ecf8-4fc2-a658-7588e2e18709)

**Comment by Secret-chest** [comment]
See #170

**Reply by WorldLanguages** [reply]
lol nice find!
three years ago
